src/spatial/utils.jl

```diff
@@ -0,0 +1,219 @@
+"""
+A file with types, structs and functions for spatial simulations
+"""
+
+"""
+stores info for a spatial jump
+"""
+struct SpatialJump{J}
+    site::J #where the jump happens
+    index::Int #diffusion or reaction
+    target_site::J #target of diffusive hop
+end
+
+############ abstract spatial system struct ##################
+"""
+Contains all info about the topology of the system
+"""
+abstract type AbstractSpatialSystem end
+
+"""
+returns neighbors of site
+"""
+function neighbors end
+
+"""
+returns total number of sites
+"""
+function number_of_sites end
+
+"""
+return the number of neighbors of a site
+"""
+function num_neighbors end
+
+# TODO make a graph struct with connectivity list, compare with CartesianGrid
+
+################### CartesianGrid <: AbstractSpatialSystem ########################
+#TODO store the number of neighbors for each site or store all neighbors for each site.
+struct CartesianGrid <: AbstractSpatialSystem
+    dimension::Int # dimension of the grid, e.g. 1, 2 or 3
+    linear_size::Int #side length of the grid
+end
+
+dimension(grid::CartesianGrid) = grid.dimension
+
+number_of_sites(grid) = grid.linear_size^dimension(grid)
+
+"""
+check if the given site_id is a valid site in the grid
+"""
+is_site(grid,site_id) = site_id >= 1 && site_id <= number_of_sites(grid)
+
+"""
+return a generator that iterates over the neighbors of the given site in grid
+"""
+function neighbors(grid, site_id)
+    (site_id + j*grid.linear_size^(i-1) for i in 1:dimension(grid), j in -1:2:1 if is_site(grid,site_id + j*grid.linear_size^(i-1)))
+end
+
+"""
+number of neighbors of the site
+"""
+function num_neighbors(grid,site_id)
+    counter = 0
+    for _ in neighbors(grid,site_id)
+        counter += 1
+    end
+    counter
+end
+
+"""
+return the nth neighbor of the site
+"""
+function nth_neighbor(grid,site,n)
+    #TODO can make this faster?
+    xs = neighbors(grid,site)
+    #TODO return negative number instead of BoundsError
+    n > 0 || throw(BoundsError(xs, n))
+
+    for (i, val) in enumerate(xs)
+        i >= n && return val
+    end
+
+    # catch iterators with no length but actual finite size less then n
+    throw(BoundsError(xs, n))
+end
+
+#TODO use the Sampler + rand interface to draw a random neighbor as described here https://docs.julialang.org/en/v1/stdlib/Random/#Random.Sampler.
+
+#TODO dealing with escaping:
+# make function that returns the ith neighbor for a site -- this might be the sink state in case of absorbing
+# or just don't update the target site if it's not a valid site
+
+################### abstract spatial rates struct ###############
+
+abstract type AbstractSpatialRates end
+
+#TODO refactor names of functions?
+
+#return rate of site
+function get_site_rate(spatial_rates_struct, site_id)
+    get_site_reactions_rate(spatial_rates_struct,site_id)+get_site_diffusions_rate(spatial_rates_struct,site_id)
+end
+
+#return rate of reactions at site
+function get_site_reactions_rate end
```

Comment:
`# return total reaction rate of the site`

similar for diffusion below

again, do these need to be forward declared or can they just be commented out and made the API for new versions?